server.go

```diff
@@ -2059,103 +2059,109 @@ func (s *server) maybeLogRecentlyAdvertisedNumEvicted() {
 	s.lastAdvertisedTxnsEvictedLogged = time.Now()
 }
 
-// handleRelayInvMsg deals with relaying inventory to peers that are not already
-// known to have it.  It is invoked from the peerHandler goroutine.
-func (s *server) handleRelayInvMsg(state *peerState, msg relayMsg) {
-	state.ForAllPeers(func(sp *serverPeer) {
-		if !sp.Connected() {
+// handleRelayPeerInvMsg handles the relay of a message to a specific
+// serverPeer.
+func (s *server) handleRelayPeerInvMsg(msg relayMsg, sp *serverPeer) {
+	if !sp.Connected() {
+		return
+	}
+
+	// Ignore peers that do not have the required service flags.
+	if !hasServices(sp.Services(), msg.reqServices) {
+		return
+	}
+
+	// Filter duplicate block announcements.
+	iv := msg.invVect
+	isBlockAnnouncement := iv.Type == wire.InvTypeBlock
+	if isBlockAnnouncement {
+		if sp.announcedBlock != nil && *sp.announcedBlock == iv.Hash {
+			sp.announcedBlock = nil
 			return
 		}
+		sp.announcedBlock = &iv.Hash
+	}
 
-		// Ignore peers that do not have the required service flags.
-		if !hasServices(sp.Services(), msg.reqServices) {
+	// Generate and send a headers message instead of an inventory message
+	// for block announcements when the peer prefers headers.
+	if isBlockAnnouncement && sp.WantsHeaders() {
+		blockHeader, ok := msg.data.(wire.BlockHeader)
+		if !ok {
+			peerLog.Warnf("Underlying data for headers" +
```

Comment:
This should fit without wrapping now due to the removed indent level.

Reply:
not quite if the goal is 80.

```
			peerLog.Warnf("Underlying data for inventory vector is not a transaction")
```

that's 98 chars wide with 8-space tabs and 86 with 4-space tabs.

Reply:
Oops, I commented on the wrong one.   I intended to comment on "Underlying data for headers is not a block header" and "Failed to add block header: %v".

Reply:
Oh, no I did comment on that one.  Hence the `blockHeader, ok` above.

Reply:
errr that's a different warnf

Reply:
I guess it probably should be a `.Warn`  and not `Warnf` too now that I look at it since there are no params.